[c10d][fr] Decouple the core logic of FR with the entry and event type

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #152614
* #152563
* __->__ #152585

We want to make FR generic enough so the first step is to make the FR a template struct so that most of common code logic can be reused. The reason for this is that CudaEvent does not inherit c10::Event and we just want to swap the event part so that for NCCL we use CudaEvent and for the rest of backends, we use c10::event.

cc @H-Huang @awgu @wanchaol @fegin @wz337 @wconstab @d4l3k

Differential Revision: [D74262695](https://our.internmc.facebook.com/intern/diff/D74262695)

fduwjj committed:
[c10d][fr] Decouple the core logic of FR with the entry and event type

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d][fr] Decouple the core logic of FR with the entry and event type"




cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

wconstab commented on torch/csrc/distributed/c10d/FlightRecorder.cpp:
```diff
@@ -20,6 +20,403 @@
 
+template <typename T, typename EntryT, typename EventType>
+std::optional<size_t> recordImpl(
+    T& obj,
```
what is 'obj'?  is this supposed to represent a ProcessGroup or something?

I wonder if we need to use Templating for T and EntryT, or we should just use templating for EventType and use polymorphism for the rest? thoughts?

fduwjj replied:
No it is a flight recorder instance so that we can differentiate between different situations (cudaEvent vs c10::Event). The signature of this function requires a Event. Actually in the next PR, it does not need this EventType.

fduwjj replied:
Also we need obj to access the members of the struct. At least for this case. The only two API we can leverage polymorphism is get_dump_json and get_dump, which I think make it templated is better? IDK since the function being called is also template not member function, so I just make them all templated.

d4l3k replied:
can we templatize this a bit tighter? I.e. have this take a FlightRecorder<T> instead of just T?

kwen2501 replied:
Similar question to why we'd need `T` and `EntryT` --
if we previously defined `class Entry`, would it just work if we extend it to `class Entry<EventType>`?

d4l3k commented on torch/csrc/distributed/c10d/FlightRecorder.cpp:
```diff
@@ -334,208 +688,35 @@ void FlightRecorder::record_accelerator_version(
-      entry->duration_ = duration;
-    }
-  }
+  retire_id_impl<FlightRecorder::Entry, Event>(
```
Why do we need to use these inner functions? Seems like we don't actually reuse them anywhere? We can just make FlightRecorder::retire_id a templatized method right?

Is the concern that we would have to move this logic into the header file to make it templatized?

d4l3k reviewed (commented):
This PR is a pretty big change -- wondering if we can keep the FR class structure and just make the existing methods template parameters rather than pulling everything out into "impl" methods

kwen2501 commented on torch/csrc/distributed/c10d/FlightRecorder.cpp:
```diff
@@ -253,65 +650,22 @@ std::optional<size_t> FlightRecorder::record(
-  }
-  return id_++;
+    bool isP2P) {
+  return recordImpl<FlightRecorder, FlightRecorder::Entry, Event>(
```
So you moved the implementation out into an independent function, calling a need to template on all classes?

Would it be easier if we keep the implementation in the class method as before, and only template the method?
```
template< typename EventType>
FlightRecorder::record(EventType* start, ...)

// At call site
FlightRecorder* fr;
fr->template record<CUDAEvent>(start, ...);  // template keyword needed here
```

fduwjj replied:
we can but the thing is that we will use that event to initiate the entry inside (which also takes in an event)..

fduwjj committed:
Update on "[c10d][fr] Decouple the core logic of FR with the entry and event type"


We want to make FR generic enough so the first step is extract common code logic into a template. The reason for this is that CudaEvent does not inherit c10::Event and there is no short work around for swapping CudaEvent with c10::event for NCCL backend, so we decide to make the common logic as a template so that in the following PR, we can enable FR for both c10d::Event and CudaEvent.


cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d][fr] Decouple the core logic of FR with the entry and event type"


We want to make FR generic enough so the first step is extract common code logic into a template. The reason for this is that CudaEvent does not inherit c10::Event and there is no short work around for swapping CudaEvent with c10::event for NCCL backend, so we decide to make the common logic as a template so that in the following PR, we can enable FR for both c10d::Event and CudaEvent.


cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

wconstab commented on torch/csrc/distributed/c10d/FlightRecorder.hpp:
```diff
@@ -188,25 +216,139 @@ struct FlightRecorder {
   // Returns the entry with the given id, if it exists. Otherwise, returns
   // std::nullopt.
-  std::optional<Entry> getEntry(std::optional<size_t> id);
+  std::optional<Entry> getEntry(std::optional<size_t> id) {
```
nit: better not to move all the code to .hpp just becuase of templating? I think you can keep functions like this in the .cpp,

and even for other functions i wonder if its worth making our own 'FlightEvent' wrapper that is templated over different event types and then just leave all the FR logic in terms of FlightEvent?

open to other ideas too

fduwjj replied:
Keeping even functions which are not dependent on Event will give link errors.

d4l3k replied:
yeah, unfortunately since it's all templatized we need it to all be in the header file.

We could potentially split this into two copies, one with just the high level API and then a second header file with "FlightRecorderImpl" to try to minimize the amount of headers included in all files. That way we only depend on the impl form the PG directly but given this is a fairly specialized interface we may be able to get that isolation regardless

For these big header-only libs I usually use folly as a prior art for how to structure. I.e. the Future class https://www.internalfb.com/code/fbsource/[6d0f557f832a72e3a7c6323a03123fab17993171]/fbcode/folly/futures/

fduwjj replied:
Somehow with the help of Ke, we can do something like this, I updated the PR and you guys can see if it makes sense.

fduwjj committed:
Update on "[c10d][fr] Decouple the core logic of FR with the entry and event type"


We want to make FR generic enough so the first step is to make the FR a template struct so that most of common code logic can be reused. The reason for this is that CudaEvent does not inherit c10::Event and we just want to swap the event part so that for NCCL we use CudaEvent and for the rest of backends, we use c10::event.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d][fr] Decouple the core logic of FR with the entry and event type"


We want to make FR generic enough so the first step is to make the FR a template struct so that most of common code logic can be reused. The reason for this is that CudaEvent does not inherit c10::Event and we just want to swap the event part so that for NCCL we use CudaEvent and for the rest of backends, we use c10::event.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

kwen2501 commented on torch/csrc/distributed/c10d/FlightRecorder.cpp:
```diff
@@ -677,6 +695,8 @@ std::string FlightRecorder::dump(
   return pickle_str(result);
 }
 
+template class FlightRecorder<at::cuda::CUDAEvent>;
+
```
Add a document here? i.e. if third party devices need to use FR, they/we would need to add
`template class FlightRecorder<c10::Event>;`
here.

fduwjj replied:
The c10::Event is added in later PR. I will comment in the later PR.

kwen2501 commented on torch/csrc/distributed/c10d/FlightRecorder.hpp:
```diff
@@ -12,6 +12,7 @@
 #include <ATen/ATen.h>
 #include <ATen/cuda/CUDAEvent.h>
 #include <c10/util/Exception.h>
+#include <nlohmann/json.hpp>
```
I guess this header is not necessary?

fduwjj replied:
Oh the next PR needs it.. let me update..

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -421,13 +421,13 @@ std::string dump_nccl_trace(
   }
 #endif // defined(USE_ROCM) && defined(NCCL_COMM_DUMP)
-  return FlightRecorder::get()->dump(
+  return FlightRecorder<at::cuda::CUDAEvent>::get()->dump(
```
nit: `using FlightRecorderCUDA = FlightRecorder<at::cuda::CUDAEvent>;`
or `typedef ...`
to reduce the verbosity.

kwen2501 reviewed (approved):
Looks much cleaner! Thanks!
Great idea templating the event types!

d4l3k reviewed (approved):
LGTM

fduwjj committed:
Update on "[c10d][fr] Decouple the core logic of FR with the entry and event type"


We want to make FR generic enough so the first step is to make the FR a template struct so that most of common code logic can be reused. The reason for this is that CudaEvent does not inherit c10::Event and we just want to swap the event part so that for NCCL we use CudaEvent and for the rest of backends, we use c10::event.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj commented:
@fduwjj has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D74262695).

fduwjj committed:
Update on "[c10d][fr] Decouple the core logic of FR with the entry and event type"


We want to make FR generic enough so the first step is to make the FR a template struct so that most of common code logic can be reused. The reason for this is that CudaEvent does not inherit c10::Event and we just want to swap the event part so that for NCCL we use CudaEvent and for the rest of backends, we use c10::event.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

Differential Revision: [D74262695](https://our.internmc.facebook.com/intern/diff/D74262695)

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d][fr] Decouple the core logic of FR with the entry and event type"


We want to make FR generic enough so the first step is to make the FR a template struct so that most of common code logic can be reused. The reason for this is that CudaEvent does not inherit c10::Event and we just want to swap the event part so that for NCCL we use CudaEvent and for the rest of backends, we use c10::event.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

Differential Revision: [D74262695](https://our.internmc.facebook.com/intern/diff/D74262695)

[ghstack-poisoned]
